src/eos-shard-shard-file-impl-v2.c

```diff
@@ -0,0 +1,259 @@
+    g_object_unref (self);
+    return NULL;
+  }
+
+  int buf_size = self->hdr.records_length * sizeof (*self->records);
+  self->records = g_malloc (buf_size);
+  pread (self->fd, self->records, buf_size, self->hdr.records_start);
+
+  return EOS_SHARD_SHARD_FILE_IMPL (self);
+}
+
+static void
+lookup_string_constant (EosShardShardFileImpl *impl, char *buf, int buf_size, uint64_t offs)
+{
+  EosShardShardFileImplV2 *self = EOS_SHARD_SHARD_FILE_IMPL_V2 (impl);
+  uint64_t global_offs = self->hdr.string_constant_table_start + offs;
+  pread (self->fd, buf, buf_size - 1, global_offs);
+}
+
+static EosShardBlob *
+blob_new (EosShardShardFileImpl *impl, struct eos_shard_v2_blob *sblob)
+{
+  EosShardShardFileImplV2 *self = EOS_SHARD_SHARD_FILE_IMPL_V2 (impl);
+  EosShardShardFile *shard_file = self->shard_file;
+
+  /* If the size is 0, then we have a blob with no data. Return null as
+   * a special case here. */
+  if (sblob->uncompressed_size == 0)
+    return NULL;
+
+  EosShardBlob *blob = _eos_shard_blob_new ();
+  blob->shard_file = g_object_ref (shard_file);
+  blob->flags = sblob->flags;
+  memcpy (blob->checksum, sblob->csum, sizeof (blob->checksum));
+  blob->size = sblob->size;
+  blob->uncompressed_size = sblob->uncompressed_size;
+  blob->offs = sblob->data_start;
+
+  char content_type[EOS_SHARD_V2_BLOB_MAX_CONTENT_TYPE_SIZE] = {};
+  lookup_string_constant (impl, content_type, sizeof (content_type), sblob->content_type_offs);
+  blob->content_type = g_strdup (content_type);
+  return blob;
+}
+
+static gboolean
+find_blob (EosShardShardFileImpl *impl, struct eos_shard_v2_record *srecord, const char *name,
+           struct eos_shard_v2_blob *blob_out)
+{
+  EosShardShardFileImplV2 *self = EOS_SHARD_SHARD_FILE_IMPL_V2 (impl);
+  int i;
+
+  /* Do a linear search because we don't expect many blobs per record... */
+  for (i = 0; i < srecord->blob_table_length; i++) {
+    struct eos_shard_v2_record_blob_table_entry entry;
+    pread (self->fd, &entry, sizeof (entry), srecord->blob_table_start + i*sizeof (entry));
+
+    struct eos_shard_v2_blob blob;
+    pread (self->fd, &blob, sizeof (blob), entry.blob_start);
+
+    char entry_name[EOS_SHARD_V2_BLOB_MAX_NAME_SIZE + 1] = {};
+    lookup_string_constant (impl, entry_name, sizeof (entry_name), blob.name_offs);
+    if (strncmp (entry_name, name, sizeof (entry_name)) == 0) {
+      *blob_out = blob;
+      return TRUE;
+    }
+  }
+
+  return FALSE;
+}
+
+static EosShardBlob *
+read_blob (EosShardShardFileImpl *impl, struct eos_shard_v2_record *srecord, const char *name)
+{
+  struct eos_shard_v2_blob blob;
+  if (!find_blob (impl, srecord, name, &blob))
+    return NULL;
+
+  return blob_new (impl, &blob);
+}
+
+static inline gboolean
+record_is_tombstone (struct eos_shard_v2_record *record)
+{
+  /* Tombstones indicate that records have been deleted.
+   * They're indicated with by a special flag in the record. */
+  return (record->flags & EOS_SHARD_V2_RECORD_FLAG_TOMBSTONE) != 0;
```

Comment:
Because that could mean it is something other than 0 or 1, which a `gboolean` should not be.